stdlib/LinearAlgebra/src/diagonal.jl

```diff
@@ -659,14 +659,19 @@ end
 # disambiguation methods: * of Diagonal and Adj/Trans AbsVec
 *(x::Adjoint{<:Any,<:AbstractVector}, D::Diagonal) = Adjoint(map((t,s) -> t'*s, D.diag, parent(x)))
 *(x::Transpose{<:Any,<:AbstractVector}, D::Diagonal) = Transpose(map((t,s) -> transpose(t)*s, D.diag, parent(x)))
-*(x::Adjoint{<:Any,<:AbstractVector}, D::Diagonal, y::AbstractVector) =
-    mapreduce(t -> t[1]*t[2]*t[3], +, zip(x, D.diag, y))
-*(x::Transpose{<:Any,<:AbstractVector}, D::Diagonal, y::AbstractVector) =
-    mapreduce(t -> t[1]*t[2]*t[3], +, zip(x, D.diag, y))
-function dot(x::AbstractVector, D::Diagonal, y::AbstractVector)
-    mapreduce(t -> dot(t[1], t[2], t[3]), +, zip(x, D.diag, y))
+*(x::Adjoint{<:Any,<:AbstractVector},   D::Diagonal, y::AbstractVector) = _dot(x, D, y)
+*(x::Transpose{<:Any,<:AbstractVector}, D::Diagonal, y::AbstractVector) = _dot(x, D, y)
+dot(x::AbstractVector, D::Diagonal, y::AbstractVector) = _dot(x, D, y)
+
+function _dot(x, D::Diagonal, y)
+    if all((isempty(x), isempty(D), isempty(y)))
```

Comment:
This reads funny. Wouldn't
```suggestion
    if isempty(x) && isempty(D) && isempty(y)
```
be easier to read and do the same job?

Reply:
Yep